Detectors/PHOS/calib/PHOSCalibWorkflow/include/PHOSCalibWorkflow/PHOSCalibCollector.h

```diff
@@ -0,0 +1,195 @@
+
+#include <TLorentzVector.h>
+#include <TH2.h>
+#include <TVector3.h>
+#include <array>
+
+using namespace o2::framework;
+
+namespace o2
+{
+namespace phos
+{
+
+// structure used to store digit info for re-calibration
+union CalibDigit {
+  uint32_t mDataWord;
+  struct {
+    uint32_t mAddress : 14;   ///< Bits  0 - 13: Hardware address
+    uint32_t mAdcAmp : 10;    ///< Bits 14 - 23: ADC counts
+    uint32_t mHgLg : 1;       ///< Bit  24: LG/HG
+    uint32_t mBadChannel : 1; ///< Bit  25: Bad channel status
+    uint32_t mCluster : 6;    ///< Bits 26-32: index of cluster in event
+  };
+};
+// Event header for energy calibraton. Allow accessing external info with vertex position and collision time
+union EventHeader {
+  uint32_t mDataWord;
+  struct {
+    uint32_t mMarker : 14; ///< Bits  0 - 13: non-existing address to separate events 16383
+    uint32_t mBC : 18;     ///< Bits 14-32: event BC (16bit in InterationRecord. Orbit (32bit) will be stored in next word)
+  };
+};
+
+// For real/mixed distribution calculation
+class RungBuffer
+{
+ public:
+  RungBuffer() = default;
+  ~RungBuffer() = default;
+
+  short size()
+  {
+    if (mFilled) {
+      return kBufferSize;
+    } else {
+      return mCurrent;
+    }
+  }
+  void addEntry(TLorentzVector& v)
+  {
+    mBuffer[mCurrent] = v;
+    mCurrent++;
+    if (mCurrent >= kBufferSize) {
+      mFilled = true;
+      mCurrent -= kBufferSize;
+    }
+  }
+  const TLorentzVector getEntry(short index)
+  {
+    //get entry from (mCurrent-1) corresponding to index=size()-1 down to size
+    if (mFilled) {
+      index += mCurrent;
+    }
+    index = index % kBufferSize;
+    return mBuffer[index];
+  }
+  //mark that next added entry will be from next event
+  void startNewEvent() { mStartCurrentEvent = mCurrent; }
+
+  bool isCurrentEvent(short index)
+  {
+    if (mCurrent >= mStartCurrentEvent) {
+      return (index >= mStartCurrentEvent && index < mCurrent);
+    } else {
+      return (index >= mStartCurrentEvent || index < mCurrent);
+    }
+  }
+
+ private:
+  static constexpr short kBufferSize = 100;        ///< Total size of the buffer
+  std::array<TLorentzVector, kBufferSize> mBuffer; ///< buffer
+  bool mFilled = false;                            ///< if buffer fully filled
+  short mCurrent = 0;                              ///< where next object will be added
+  short mStartCurrentEvent = 0;                    ///< start of current event
+};
+
+class PHOSCalibCollector : public o2::framework::Task
+{
+
+  //Histogram kinds to be filled
+  enum hnames { kReInvMassPerCell,
+                kMiInvMassPerCell,
+                kReInvMassNonlin,
+                kMiInvMassNonlin,
+                kTimeHGPerCell,
+                kTimeLGPerCell,
+                kTimeHGSlewing,
+                kTimeLGSlewing };
+
+ public:
+  PHOSCalibCollector() = default;
+  PHOSCalibCollector(short mode) : mMode(mode) {}
+
+  ~PHOSCalibCollector() final = default;
+
```

Comment:
Dear @shahor02 ,
what code checked suggests here? Initial version with ~PHOSCalibCollector() = default; checker criticized and suggested adding final/override. This version also causes warning. I think, this should be final class and I do not want to write descructor explicitely. What should I write here? Thanks, Dmitri

Reply:
@peressounko the LLVM produces a warning (which we interpret as an error) if the class having a virtual destructor and not flagged as ``final`` declares its destructor as ``final``. Use ``override = default`` instead.